CRM/Core/Payment/Dummy.php

```diff
@@ -137,6 +137,38 @@ protected function supportsLiveMode() {
     return TRUE;
   }
 
+  /**
+   * Does this payment processor support refund?
+   *
+   * @return bool
+   */
+  public function supportsRefund() {
+    return TRUE;
+  }
+
+  /**
+   * Submit a refund payment using Advanced Integration Method.
+   *
+   * @param array $params
+   *   Assoc array of input parameters for this transaction.
+   *
+   * @return array
+   *   the result in a nice formatted array (or an error object)
+   */
+  public function doRefundPayment(&$params) {
+    if (empty($params['trxn_id'])) {
+      $error = new CRM_Core_Error(ts('Refund failed'));
```

Comment:
what are we trying to achieve here?

Reply:
@eileenmcnaughton ideally all PP to do a refund we must have trxn_id against with a refund payment is claimed. However, there might be additional parameters needed and will vary for different PP. Code here tries to mimic the behavior for dummy processor to ensure that trxn_id is always present. 

This exception can be avoided if we ensure at form or at API level trxn_id is a mandatory field.